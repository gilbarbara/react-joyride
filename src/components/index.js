import React from 'react';
import PropTypes from 'prop-types';
import treeChanges from 'tree-changes';
import is from 'is-lite';

import Store from '../modules/store';
import {
  getElement,
  getScrollTo,
  getScrollParent,
  hasCustomScrollParent,
  isFixed,
  scrollTo,
} from '../modules/dom';
import { canUseDOM, isEqual, log } from '../modules/helpers';
import { getMergedStep, validateSteps } from '../modules/step';

import ACTIONS from '../constants/actions';
import EVENTS from '../constants/events';
import LIFECYCLE from '../constants/lifecycle';
import STATUS from '../constants/status';

import Step from './Step';

class Joyride extends React.Component {
  constructor(props) {
    super(props);

    this.store = new Store({
      ...props,
      controlled: props.run && is.number(props.stepIndex),
    });
    this.state = this.store.getState();
    this.helpers = this.store.getHelpers();
  }

  static propTypes = {
    beaconComponent: PropTypes.oneOfType([
      PropTypes.func,
      PropTypes.element,
    ]),
    callback: PropTypes.func,
    continuous: PropTypes.bool,
    debug: PropTypes.bool,
    disableCloseOnEsc: PropTypes.bool,
    disableOverlay: PropTypes.bool,
    disableOverlayClose: PropTypes.bool,
    disableScrolling: PropTypes.bool,
    floaterProps: PropTypes.shape({
      offset: PropTypes.number,
    }),
    hideBackButton: PropTypes.bool,
    locale: PropTypes.object,
    run: PropTypes.bool,
    scrollOffset: PropTypes.number,
    scrollToFirstStep: PropTypes.bool,
    showProgress: PropTypes.bool,
    showSkipButton: PropTypes.bool,
    spotlightClicks: PropTypes.bool,
    spotlightPadding: PropTypes.number,
    stepIndex: PropTypes.number,
    steps: PropTypes.array,
    styles: PropTypes.object,
    tooltipComponent: PropTypes.oneOfType([
      PropTypes.func,
      PropTypes.element,
    ]),
  };

  static defaultProps = {
    continuous: false,
    debug: false,
    disableCloseOnEsc: false,
    disableOverlay: false,
    disableOverlayClose: false,
    disableScrolling: false,
    hideBackButton: false,
    run: true,
    scrollOffset: 20,
    scrollToFirstStep: false,
    showSkipButton: false,
    showProgress: false,
    spotlightClicks: false,
    spotlightPadding: 10,
    steps: [],
  };

  componentDidMount() {
    if (!canUseDOM) return;

    const {
      debug,
      disableCloseOnEsc,
      run,
      steps,
    } = this.props;
    const { start } = this.store;

    log({
      title: 'init',
      data: [
        { key: 'props', value: this.props },
        { key: 'state', value: this.state },
      ],
      debug,
    });

    // Sync the store to this component state.
    this.store.addListener(this.syncState);

    if (validateSteps(steps, debug) && run) {
      start();
    }

    /* istanbul ignore else */
    if (!disableCloseOnEsc) {
      document.body.addEventListener('keydown', this.handleKeyboard, { passive: true });
    }
  }

  componentWillReceiveProps(nextProps) {
    if (!canUseDOM) return;
    const { action, status } = this.state;
    const { steps, stepIndex } = this.props;
    const { debug, run, steps: nextSteps, stepIndex: nextStepIndex } = nextProps;
    const { setSteps, start, stop, update } = this.store;
    const diffProps = !isEqual(this.props, nextProps);
    const { changed } = treeChanges(this.props, nextProps);

    if (diffProps) {
      log({
        title: 'props',
        data: [
          { key: 'nextProps', value: nextProps },
          { key: 'props', value: this.props },
        ],
        debug,
      });

      const stepsChanged = !isEqual(nextSteps, steps);
      const stepIndexChanged = is.number(nextStepIndex) && changed('stepIndex');

      /* istanbul ignore else */
      if (changed('run')) {
        if (run) {
          start(nextStepIndex);
        }
        else {
          stop();
        }
      }

      if (stepsChanged) {
        if (validateSteps(nextSteps, debug)) {
          setSteps(nextSteps);
        }
        else {
          console.warn('Steps are not valid', nextSteps); //eslint-disable-line no-console
        }
      }

      /* istanbul ignore else */
      if (stepIndexChanged) {
        let nextAction = stepIndex < nextStepIndex ? ACTIONS.NEXT : ACTIONS.PREV;

        if (action === ACTIONS.STOP) {
          nextAction = ACTIONS.START;
        }

        if (![STATUS.FINISHED, STATUS.SKIPPED].includes(status)) {
          update({
            action: action === ACTIONS.CLOSE ? ACTIONS.CLOSE : nextAction,
            index: nextStepIndex,
            lifecycle: LIFECYCLE.INIT,
          });
        }
      }
    }
  }

  componentDidUpdate(prevProps, prevState) {
    if (!canUseDOM) return;

    const { index, lifecycle, status } = this.state;
    const { debug, steps } = this.props;
    const { changed, changedFrom, changedTo } = treeChanges(prevState, this.state);
    const diffState = !isEqual(prevState, this.state);
    let step = getMergedStep(steps[index], this.props);

    if (diffState) {
      log({
        title: 'state',
        data: [
          { key: 'state', value: this.state },
          { key: 'changed', value: diffState },
          { key: 'step', value: step },
        ],
        debug,
      });

      let currentIndex = index;

      if (changed('status')) {
        let type = EVENTS.TOUR_STATUS;

        if (changedTo('status', STATUS.FINISHED) || changedTo('status', STATUS.SKIPPED)) {
          type = EVENTS.TOUR_END;
          // Return the last step when the tour is finished
          step = getMergedStep(steps[prevState.index], this.props);
          currentIndex = prevState.index;
        }
        else if (changedFrom('status', STATUS.READY, STATUS.RUNNING)) {
          type = EVENTS.TOUR_START;
        }

        this.callback({
          ...this.state,
          index: currentIndex,
          step,
          type,
        });
      }

      if (step) {
        this.scrollToStep(prevState);

        if (step.placement === 'center' && status === STATUS.RUNNING && lifecycle === LIFECYCLE.INIT) {
          this.store.update({ lifecycle: LIFECYCLE.READY });
        }
      }

      if (changedTo('lifecycle', LIFECYCLE.INIT)) {
        delete this.beaconPopper;
        delete this.tooltipPopper;
      }
    }
  }

  componentWillUnmount() {
    const { disableCloseOnEsc } = this.props;

    /* istanbul ignore else */
    if (!disableCloseOnEsc) {
      document.body.removeEventListener('keydown', this.handleKeyboard);
    }
  }

  scrollToStep(prevState) {
    const { index, lifecycle, status } = this.state;
    const { debug, disableScrolling, scrollToFirstStep, scrollOffset, steps } = this.props;
    const step = getMergedStep(steps[index], this.props);

    if (step) {
      const target = getElement(step.target);

      const shouldScroll = step
        && !disableScrolling
        && (!step.isFixed || !isFixed(target)) // fixed steps don't need to scroll
        && (prevState.lifecycle !== lifecycle && [LIFECYCLE.BEACON, LIFECYCLE.TOOLTIP].includes(lifecycle))
        && (scrollToFirstStep || prevState.index !== index);

      if (status === STATUS.RUNNING && shouldScroll) {
        const hasCustomScroll = hasCustomScrollParent(target);
        const scrollParent = getScrollParent(target);
        let scrollY = Math.floor(getScrollTo(target, scrollOffset));

        log({
          title: 'scrollToStep',
          data: [
            { key: 'index', value: index },
            { key: 'lifecycle', value: lifecycle },
            { key: 'status', value: status },
          ],
          debug,
        });

        if (lifecycle === LIFECYCLE.BEACON && this.beaconPopper) {
          const { placement, popper } = this.beaconPopper;

          if (!['bottom'].includes(placement) && !hasCustomScroll) {
            scrollY = Math.floor(popper.top - scrollOffset);
          }
        }
        else if (lifecycle === LIFECYCLE.TOOLTIP && this.tooltipPopper) {
          const { flipped, placement, popper } = this.tooltipPopper;

          if (['top', 'right'].includes(placement) && !flipped && !hasCustomScroll) {
            scrollY = Math.floor(popper.top - scrollOffset);
          }
          else {
            scrollY -= step.spotlightPadding;
          }
        }

        if (status === STATUS.RUNNING && shouldScroll && scrollY >= 0) {
          scrollTo(scrollY, scrollParent);
        }
      }
    }
  }

  /**
   * Trigger the callback.
   *
   * @private
   * @param {Object} data
   */
  callback = (data) => {
    const { callback } = this.props;

    /* istanbul ignore else */
    if (is.function(callback)) {
      callback(data);
    }
  };

  /**
   * Keydown event listener
   *
   * @private
   * @param {Event} e - Keyboard event
   */
  handleKeyboard = (e) => {
    const { index, lifecycle } = this.state;
    const { steps } = this.props;
    const step = steps[index];
    const intKey = window.Event ? e.which : e.keyCode;

    if (lifecycle === LIFECYCLE.TOOLTIP) {
      if (intKey === 27 && (step && !step.disableCloseOnEsc)) {
        this.store.close();
      }
    }
  };

  /**
   * Sync the store with the component's state
   *
   * @param {Object} state
   */
  syncState = (state) => {
    this.setState(state);
  };

  getPopper = (popper, type) => {
    if (type === 'wrapper') {
      this.beaconPopper = popper;
    }
    else {
      this.tooltipPopper = popper;
    }
  };

  render() {
    if (!canUseDOM) return null;

    const { index, status } = this.state;
    const { continuous, debug, disableScrolling, steps } = this.props;
    const step = getMergedStep(steps[index], this.props);
    let output;

    if (status === STATUS.RUNNING && step) {
      output = (
        <Step
          {...this.state}
          callback={this.callback}
          continuous={continuous}
          debug={debug}
          disableScrolling={disableScrolling}
          getPopper={this.getPopper}
          helpers={this.helpers}
          step={step}
          update={this.store.update}
        />
      );
    }

    return (
      <div className="joyride">
        {output}
      </div>
    );
  }
}

export default Joyride;
