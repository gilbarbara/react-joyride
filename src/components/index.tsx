import * as React from 'react';
import isEqual from '@gilbarbara/deep-equal';
import is from 'is-lite';
import treeChanges from 'tree-changes';

import {
  canUseDOM,
  getElement,
  getScrollParent,
  getScrollTo,
  hasCustomScrollParent,
  scrollTo,
} from '~/modules/dom';
import { log, shouldScroll } from '~/modules/helpers';
import { getMergedStep, validateSteps } from '~/modules/step';
import createStore from '~/modules/store';

import { ACTIONS, EVENTS, LIFECYCLE, STATUS } from '~/literals';

import { defaultProps } from '~/defaults';
import { Actions, CallBackProps, Props, State, Status, StoreHelpers } from '~/types';

import Step from './Step';

class Joyride extends React.Component<Props, State> {
  private readonly helpers: StoreHelpers;
  private readonly store: ReturnType<typeof createStore>;

  static defaultProps = defaultProps;

  constructor(props: Props) {
    super(props);

    const { debug, getHelpers, run, stepIndex } = props;

    this.store = createStore({
      ...props,
      controlled: run && is.number(stepIndex),
    });
    this.helpers = this.store.getHelpers();

    const { addListener } = this.store;

    log({
      title: 'init',
      data: [
        { key: 'props', value: this.props },
        { key: 'state', value: this.state },
      ],
      debug,
    });

    // Sync the store to this component's state.
    addListener(this.syncState);

    if (getHelpers) {
      getHelpers(this.helpers);
    }

    this.state = this.store.getState();
  }

  componentDidMount() {
    if (!canUseDOM()) {
      return;
    }

    const { debug, disableCloseOnEsc, run, steps } = this.props;
    const { start } = this.store;

    if (validateSteps(steps, debug) && run) {
      start();
    }

    /* istanbul ignore else */
    if (!disableCloseOnEsc) {
      document.body.addEventListener('keydown', this.handleKeyboard, { passive: true });
    }
  }

  componentDidUpdate(previousProps: Props, previousState: State) {
    if (!canUseDOM()) {
      return;
    }

    const { action, controlled, index, lifecycle, status } = this.state;
    const { debug, run, stepIndex, steps } = this.props;
    const { stepIndex: previousStepIndex, steps: previousSteps } = previousProps;
    const { reset, setSteps, start, stop, update } = this.store;
    const { changed: changedProps } = treeChanges(previousProps, this.props);
    const { changed, changedFrom } = treeChanges(previousState, this.state);
    const step = getMergedStep(steps[index], this.props);

    const stepsChanged = !isEqual(previousSteps, steps);
    const stepIndexChanged = is.number(stepIndex) && changedProps('stepIndex');
    const target = getElement(step.target);

    if (stepsChanged) {
      if (validateSteps(steps, debug)) {
        setSteps(steps);
      } else {
        // eslint-disable-next-line no-console
        console.warn('Steps are not valid', steps);
      }
    }

    /* istanbul ignore else */
    if (changedProps('run')) {
      if (run) {
        start(stepIndex);
      } else {
        stop();
      }
    }

    /* istanbul ignore else */
    if (stepIndexChanged) {
      let nextAction: Actions =
        is.number(previousStepIndex) && previousStepIndex < stepIndex ? ACTIONS.NEXT : ACTIONS.PREV;

      if (action === ACTIONS.STOP) {
        nextAction = ACTIONS.START;
      }

      if (!([STATUS.FINISHED, STATUS.SKIPPED] as Array<Status>).includes(status)) {
        update({
          action: action === ACTIONS.CLOSE ? ACTIONS.CLOSE : nextAction,
          index: stepIndex,
          lifecycle: LIFECYCLE.INIT,
        });
      }
    }

    // Update the index if the first step is not found
    if (!controlled && status === STATUS.RUNNING && index === 0 && !target) {
      this.store.update({ index: index + 1 });
      this.callback({
        ...this.state,
        type: EVENTS.TARGET_NOT_FOUND,
        step,
      });
    }

    const callbackData = {
      ...this.state,
      index,
      step,
    };
    const isAfterAction = changed('action', [
      ACTIONS.NEXT,
      ACTIONS.PREV,
      ACTIONS.SKIP,
      ACTIONS.CLOSE,
    ]);

    if (isAfterAction && changed('status', STATUS.PAUSED)) {
      const previousStep = getMergedStep(steps[previousState.index], this.props);

      this.callback({
        ...callbackData,
        index: previousState.index,
        lifecycle: LIFECYCLE.COMPLETE,
        step: previousStep,
        type: EVENTS.STEP_AFTER,
      });
    }

    if (changed('status', [STATUS.FINISHED, STATUS.SKIPPED])) {
      const previousStep = getMergedStep(steps[previousState.index], this.props);

      if (!controlled) {
        this.callback({
          ...callbackData,
          index: previousState.index,
          lifecycle: LIFECYCLE.COMPLETE,
          step: previousStep,
          type: EVENTS.STEP_AFTER,
        });
      }

      this.callback({
        ...callbackData,
        type: EVENTS.TOUR_END,
        // Return the last step when the tour is finished
        step: previousStep,
        index: previousState.index,
      });
      reset();
    } else if (changedFrom('status', [STATUS.IDLE, STATUS.READY], STATUS.RUNNING)) {
      this.callback({
        ...callbackData,
        type: EVENTS.TOUR_START,
      });
    } else if (changed('status') || changed('action', ACTIONS.RESET)) {
      this.callback({
        ...callbackData,
        type: EVENTS.TOUR_STATUS,
      });
    }

    this.scrollToStep(previousState);

    if (step.placement === 'center' && status === STATUS.RUNNING && lifecycle === LIFECYCLE.INIT) {
      this.store.update({ lifecycle: LIFECYCLE.READY });
    }
  }

  componentWillUnmount() {
    const { disableCloseOnEsc } = this.props;

    /* istanbul ignore else */
    if (!disableCloseOnEsc) {
      document.body.removeEventListener('keydown', this.handleKeyboard);
    }
  }

  /**
   * Trigger the callback.
   */
  callback = (data: CallBackProps) => {
    const { callback } = this.props;

    /* istanbul ignore else */
    if (is.function(callback)) {
      callback(data);
    }
  };

  /**
   * Keydown event listener
   */
  handleKeyboard = (event: KeyboardEvent) => {
    const { index, lifecycle } = this.state;
    const { steps } = this.props;
    const step = steps[index];

    if (lifecycle === LIFECYCLE.TOOLTIP) {
      if (event.code === 'Escape' && step && !step.disableCloseOnEsc) {
        this.store.close('keyboard');
      }
    }
  };

  /**
   * Sync the store with the component's state
   */
  syncState = (state: State) => {
    this.setState(state);
  };

  scrollToStep(previousState: State) {
    const { index, lifecycle, status } = this.state;
    const {
      debug,
      disableScrollParentFix = false,
      scrollDuration,
      scrollOffset = 20,
      scrollToFirstStep = false,
      steps,
    } = this.props;
    const step = getMergedStep(steps[index], this.props);

    const target = getElement(step.target);
    const shouldScrollToStep = shouldScroll({
      isFirstStep: index === 0,
      lifecycle,
      previousLifecycle: previousState.lifecycle,
      scrollToFirstStep,
      step,
      target,
    });

    if (status === STATUS.RUNNING && shouldScrollToStep) {
      const hasCustomScroll = hasCustomScrollParent(target, disableScrollParentFix);
      const scrollParent = getScrollParent(target, disableScrollParentFix);
      let scrollY = Math.floor(getScrollTo(target, scrollOffset, disableScrollParentFix)) || 0;

      log({
        title: 'scrollToStep',
        data: [
          { key: 'index', value: index },
          { key: 'lifecycle', value: lifecycle },
          { key: 'status', value: status },
        ],
        debug,
      });

      const beaconPopper = this.store.getPopper('beacon');
      const tooltipPopper = this.store.getPopper('tooltip');

      /* istanbul ignore else */
      if (lifecycle === LIFECYCLE.BEACON && beaconPopper) {
        const { offsets, placement } = beaconPopper;

        /* istanbul ignore else */
        if (!['bottom'].includes(placement) && !hasCustomScroll) {
          scrollY = Math.floor(offsets.popper.top - scrollOffset);
        }
      } else if (lifecycle === LIFECYCLE.TOOLTIP && tooltipPopper) {
        const { flipped, offsets, placement } = tooltipPopper;

        if (['top', 'right', 'left'].includes(placement) && !flipped && !hasCustomScroll) {
          scrollY = Math.floor(offsets.popper.top - scrollOffset);
        } else {
          scrollY -= step.spotlightPadding;
        }
      }

      scrollY = scrollY >= 0 ? scrollY : 0;

      /* istanbul ignore else */
      if (status === STATUS.RUNNING) {
        scrollTo(scrollY, { element: scrollParent as Element, duration: scrollDuration }).then(
          () => {
            setTimeout(() => {
              this.store.getPopper('tooltip')?.instance.update();
            }, 10);
          },
        );
      }
    }
  }

  render() {
    if (!canUseDOM()) {
      return null;
    }

    const { index, status } = this.state;
    const {
      continuous = false,
      debug = false,
      nonce,
      scrollToFirstStep = false,
      steps,
    } = this.props;
    let output;

    if (status === STATUS.RUNNING && steps[index]) {
      const step = getMergedStep(steps[index], this.props);

      output = (
        <Step
          {...this.state}
          callback={this.callback}
          continuous={continuous}
          debug={debug}
          helpers={this.helpers}
          nonce={nonce}
          shouldScroll={!step.disableScrolling && (index !== 0 || scrollToFirstStep)}
          step={step}
          store={this.store}
        />
      );
    }

    return <div className="react-joyride">{output}</div>;
  }
}

export default Joyride;
