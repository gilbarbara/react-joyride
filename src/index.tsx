export * from './literals';

// eslint-disable-next-line no-restricted-exports
export { default } from './components';
export * from './types';
