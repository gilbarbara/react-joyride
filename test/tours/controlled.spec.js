import React from 'react';

import Controlled from '../__fixtures__/Controlled';

import { ACTIONS, EVENTS, LIFECYCLE, STATUS } from '../../src';

jest.mock('popper.js', () => {
  const PopperJS = jest.requireActual('popper.js');

  return class {
    static placements = PopperJS.placements;

    constructor() {
      return {
        destroy: () => {},
        scheduleUpdate: () => {},
      };
    }
  };
});

console.warn = jest.fn();

const mockCallback = jest.fn();
const props = {
  callback: mockCallback,
};

function initPopper(wrapper) {
  const Step = wrapper.find('JoyrideStep').instance();
  Step.setPopper({ popper: {} }, 'wrapper');
  Step.setPopper({ popper: {} }, 'tooltip');
}

describe('Joyride > Controlled', () => {
  let wrapper;
  let joyride;

  beforeAll(() => {
    wrapper = mount(<Controlled {...props} />, { attachTo: document.getElementById('react') });
    joyride = wrapper.find('Joyride').instance();
  });

  it('should have initiated the Joyride', () => {
    expect(joyride.state).toMatchSnapshot();
  });

  it('should not have rendered a step', () => {
    expect(wrapper.find('Joyride')).toExist();
    expect(wrapper.find('JoyrideStep')).not.toExist();
  });

  it('should be able to start the tour', () => {
    wrapper.find('.hero__start').simulate('click');

    expect(joyride.state).toMatchSnapshot();
    expect(mockCallback).toHaveBeenCalledWith({
      action: ACTIONS.START,
      controlled: true,
      index: 0,
      lifecycle: LIFECYCLE.INIT,
      size: 5,
      status: STATUS.RUNNING,
      step: expect.any(Object),
      type: EVENTS.TOUR_START,
    });
  });

  it('should have rendered the STEP 1 Beacon', () => {
    initPopper(wrapper);

    expect(mockCallback).toHaveBeenCalledWith({
      action: ACTIONS.START,
      controlled: true,
      index: 0,
      lifecycle: LIFECYCLE.READY,
      size: 5,
      status: STATUS.RUNNING,
      step: expect.any(Object),
      type: EVENTS.STEP_BEFORE,
    });

    expect(mockCallback).toHaveBeenCalledWith({
      action: ACTIONS.UPDATE,
      controlled: true,
      index: 0,
      lifecycle: LIFECYCLE.BEACON,
      size: 5,
      status: STATUS.RUNNING,
      step: expect.any(Object),
      type: EVENTS.BEACON,
    });
  });

  it('should be able to click STEP 1 Beacon', () => {
    wrapper.find('JoyrideBeacon').simulate('click');

    expect(mockCallback).toHaveBeenCalledWith({
      action: ACTIONS.UPDATE,
      controlled: true,
      index: 0,
      lifecycle: LIFECYCLE.TOOLTIP,
      size: 5,
      status: STATUS.RUNNING,
      step: expect.any(Object),
      type: EVENTS.TOOLTIP,
    });
    expect(joyride.state).toMatchSnapshot();
  });

  it('should be able to click STEP 1 Primary button', () => {
    wrapper.find('JoyrideTooltip [data-test-id="button-primary"]').simulate('click');

    expect(mockCallback).toHaveBeenCalledWith({
      action: ACTIONS.NEXT,
      controlled: true,
      index: 0,
      lifecycle: LIFECYCLE.COMPLETE,
      size: 5,
      status: STATUS.RUNNING,
      step: expect.any(Object),
      type: EVENTS.STEP_AFTER,
    });
    expect(joyride.state).toMatchSnapshot();
  });

  it('should have rendered the STEP 2 Tooltip', () => {
    initPopper(wrapper);

    expect(mockCallback).toHaveBeenCalledWith({
      action: ACTIONS.NEXT,
      controlled: true,
      index: 1,
      lifecycle: LIFECYCLE.READY,
      size: 5,
      status: STATUS.RUNNING,
      step: expect.any(Object),
      type: EVENTS.STEP_BEFORE,
    });

    expect(mockCallback).toHaveBeenCalledWith({
      action: ACTIONS.UPDATE,
      controlled: true,
      index: 1,
      lifecycle: LIFECYCLE.TOOLTIP,
      size: 5,
      status: STATUS.RUNNING,
      step: expect.any(Object),
      type: EVENTS.TOOLTIP,
    });
  });

  it('should be able to click STEP 2 button through the spotlight', () => {
    wrapper.find('.mission__button').simulate('click');

    expect(mockCallback).toHaveBeenCalledWith({
      action: ACTIONS.NEXT,
      controlled: true,
      index: 1,
      lifecycle: LIFECYCLE.COMPLETE,
      size: 5,
      status: STATUS.RUNNING,
      step: expect.any(Object),
      type: EVENTS.STEP_AFTER,
    });
    expect(joyride.state).toMatchSnapshot();
  });

  it('should have rendered the STEP 3 Tooltip', () => {
    initPopper(wrapper);

    expect(mockCallback).toHaveBeenCalledWith({
      action: ACTIONS.NEXT,
      controlled: true,
      index: 2,
      lifecycle: LIFECYCLE.READY,
      size: 5,
      status: STATUS.RUNNING,
      step: expect.any(Object),
      type: EVENTS.STEP_BEFORE,
    });

    expect(mockCallback).toHaveBeenCalledWith({
      action: ACTIONS.UPDATE,
      controlled: true,
      index: 2,
      lifecycle: LIFECYCLE.TOOLTIP,
      size: 5,
      status: STATUS.RUNNING,
      step: expect.any(Object),
      type: EVENTS.TOOLTIP,
    });
    expect(joyride.state).toMatchSnapshot();
  });

  it('should be able to click STEP 3 Back button', () => {
    wrapper.find('JoyrideTooltip [data-test-id="button-back"]').simulate('click');

    expect(mockCallback).toHaveBeenCalledWith({
      action: ACTIONS.PREV,
      controlled: true,
      index: 2,
      lifecycle: LIFECYCLE.COMPLETE,
      size: 5,
      status: STATUS.RUNNING,
      step: expect.any(Object),
      type: EVENTS.STEP_AFTER,
    });
    expect(joyride.state).toMatchSnapshot();
  });

  it('should have rendered the STEP 2 Tooltip and advance AGAIN', () => {
    initPopper(wrapper);

    expect(mockCallback).toHaveBeenCalledWith({
      action: ACTIONS.PREV,
      controlled: true,
      index: 1,
      lifecycle: LIFECYCLE.READY,
      size: 5,
      status: STATUS.RUNNING,
      step: expect.any(Object),
      type: EVENTS.STEP_BEFORE,
    });

    expect(mockCallback).toHaveBeenCalledWith({
      action: ACTIONS.UPDATE,
      controlled: true,
      index: 1,
      lifecycle: LIFECYCLE.TOOLTIP,
      size: 5,
      status: STATUS.RUNNING,
      step: expect.any(Object),
      type: EVENTS.TOOLTIP,
    });
  });

  it('should be able to click STEP 2 button through the spotlight AGAIN', () => {
    wrapper.find('.mission__button').simulate('click');

    expect(mockCallback).toHaveBeenCalledWith({
      action: ACTIONS.NEXT,
      controlled: true,
      index: 1,
      lifecycle: LIFECYCLE.COMPLETE,
      size: 5,
      status: STATUS.RUNNING,
      step: expect.any(Object),
      type: EVENTS.STEP_AFTER,
    });
    expect(joyride.state).toMatchSnapshot();
  });

  it('should have rendered the STEP 3 Tooltip AGAIN', () => {
    initPopper(wrapper);

    expect(mockCallback).toHaveBeenCalledWith({
      action: ACTIONS.NEXT,
      controlled: true,
      index: 2,
      lifecycle: LIFECYCLE.READY,
      size: 5,
      status: STATUS.RUNNING,
      step: expect.any(Object),
      type: EVENTS.STEP_BEFORE,
    });

    expect(mockCallback).toHaveBeenCalledWith({
      action: ACTIONS.UPDATE,
      controlled: true,
      index: 2,
      lifecycle: LIFECYCLE.TOOLTIP,
      size: 5,
      status: STATUS.RUNNING,
      step: expect.any(Object),
      type: EVENTS.TOOLTIP,
    });
    expect(joyride.state).toMatchSnapshot();
  });

  it('should be able to click STEP 3 Primary button', () => {
    wrapper.find('JoyrideTooltip [data-test-id="button-primary"]').simulate('click');

    expect(mockCallback).toHaveBeenCalledWith({
      action: ACTIONS.NEXT,
      controlled: true,
      index: 2,
      lifecycle: LIFECYCLE.COMPLETE,
      size: 5,
      status: STATUS.RUNNING,
      step: expect.any(Object),
      type: EVENTS.STEP_AFTER,
    });
    expect(joyride.state).toMatchSnapshot();
  });

  it('should have skipped the STEP 4', () => {
    expect(console.warn).toHaveBeenCalledTimes(1);
    expect(mockCallback).toHaveBeenCalledWith({
      action: ACTIONS.NEXT,
      controlled: true,
      index: 3,
      lifecycle: LIFECYCLE.INIT,
      size: 5,
      status: STATUS.RUNNING,
      step: expect.any(Object),
      type: EVENTS.TARGET_NOT_FOUND,
    });
  });

  it('should have rendered the STEP 5 Tooltip', () => {
    initPopper(wrapper);

    expect(mockCallback).toHaveBeenCalledWith({
      action: ACTIONS.NEXT,
      controlled: true,
      index: 4,
      lifecycle: LIFECYCLE.READY,
      size: 5,
      status: STATUS.RUNNING,
      step: expect.any(Object),
      type: EVENTS.STEP_BEFORE,
    });

    expect(mockCallback).toHaveBeenCalledWith({
      action: ACTIONS.UPDATE,
      controlled: true,
      index: 4,
      lifecycle: LIFECYCLE.TOOLTIP,
      size: 5,
      status: STATUS.RUNNING,
      step: expect.any(Object),
      type: EVENTS.TOOLTIP,
    });
    expect(joyride.state).toMatchSnapshot();
  });

  it('should be able to click STEP 5 Primary button', () => {
    wrapper.find('JoyrideTooltip [data-test-id="button-primary"]').simulate('click');

    expect(mockCallback).toHaveBeenCalledWith({
      action: ACTIONS.NEXT,
      controlled: true,
      index: 4,
      lifecycle: LIFECYCLE.COMPLETE,
      size: 5,
      status: STATUS.RUNNING,
      step: expect.any(Object),
      type: EVENTS.STEP_AFTER,
    });
    expect(joyride.state).toMatchSnapshot();
  });

  it('should have ended the tour', () => {
    expect(mockCallback).toHaveBeenCalledWith({
      action: ACTIONS.NEXT,
      controlled: true,
      index: 4,
      lifecycle: LIFECYCLE.INIT,
      size: 5,
      status: STATUS.FINISHED,
      step: expect.any(Object),
      type: EVENTS.TOUR_END,
    });
  });

  it('should unmount', () => {
    wrapper.unmount();

    expect(wrapper.find('Joyride')).not.toExist();
  });
});
