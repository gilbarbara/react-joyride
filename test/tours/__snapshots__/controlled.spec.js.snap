// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Joyride > Controlled should be able to click STEP 1 Primary button 1`] = `
{
  "action": "next",
  "controlled": true,
  "index": 1,
  "lifecycle": "init",
  "size": 7,
  "status": "running",
}
`;

exports[`Joyride > Controlled should be able to click STEP 2 Primary button 1`] = `
{
  "action": "next",
  "controlled": true,
  "index": 2,
  "lifecycle": "init",
  "size": 7,
  "status": "running",
}
`;

exports[`Joyride > Controlled should be able to click STEP 3 button through the spotlight 1`] = `
{
  "action": "next",
  "controlled": true,
  "index": 3,
  "lifecycle": "init",
  "size": 7,
  "status": "running",
}
`;

exports[`Joyride > Controlled should be able to click STEP 3 button through the spotlight AGAIN 1`] = `
{
  "action": "next",
  "controlled": true,
  "index": 3,
  "lifecycle": "init",
  "size": 7,
  "status": "running",
}
`;

exports[`Joyride > Controlled should be able to click STEP 4 Back button 1`] = `
{
  "action": "prev",
  "controlled": true,
  "index": 2,
  "lifecycle": "init",
  "size": 7,
  "status": "running",
}
`;

exports[`Joyride > Controlled should be able to click STEP 4 Primary button 1`] = `
{
  "action": "next",
  "controlled": true,
  "index": 5,
  "lifecycle": "init",
  "size": 7,
  "status": "running",
}
`;

exports[`Joyride > Controlled should be able to click STEP 6 Primary button 1`] = `
{
  "action": "update",
  "controlled": true,
  "index": 6,
  "lifecycle": "tooltip",
  "size": 7,
  "status": "running",
}
`;

exports[`Joyride > Controlled should be able to click STEP 7 Primary button and end the tour 1`] = `
{
  "action": "next",
  "controlled": true,
  "index": 7,
  "lifecycle": "init",
  "size": 7,
  "status": "finished",
}
`;

exports[`Joyride > Controlled should be able to start the tour and render the centered tooltip 1`] = `
{
  "action": "update",
  "controlled": true,
  "index": 0,
  "lifecycle": "tooltip",
  "size": 7,
  "status": "running",
}
`;

exports[`Joyride > Controlled should have initiated the Joyride 1`] = `
{
  "action": "init",
  "controlled": true,
  "index": 0,
  "lifecycle": "init",
  "size": 7,
  "status": "ready",
}
`;

exports[`Joyride > Controlled should have rendered the STEP 2 Tooltip 1`] = `
{
  "action": "update",
  "controlled": true,
  "index": 1,
  "lifecycle": "tooltip",
  "size": 7,
  "status": "running",
}
`;

exports[`Joyride > Controlled should have rendered the STEP 3 Tooltip 1`] = `
{
  "action": "update",
  "controlled": true,
  "index": 2,
  "lifecycle": "tooltip",
  "size": 7,
  "status": "running",
}
`;

exports[`Joyride > Controlled should have rendered the STEP 3 Tooltip and advance AGAIN 1`] = `
{
  "action": "update",
  "controlled": true,
  "index": 2,
  "lifecycle": "tooltip",
  "size": 7,
  "status": "running",
}
`;

exports[`Joyride > Controlled should have rendered the STEP 4 Tooltip 1`] = `
{
  "action": "update",
  "controlled": true,
  "index": 3,
  "lifecycle": "tooltip",
  "size": 7,
  "status": "running",
}
`;

exports[`Joyride > Controlled should have rendered the STEP 4 Tooltip AGAIN 1`] = `
{
  "action": "update",
  "controlled": true,
  "index": 3,
  "lifecycle": "tooltip",
  "size": 7,
  "status": "running",
}
`;

exports[`Joyride > Controlled should have rendered the STEP 6 Tooltip 1`] = `
{
  "action": "update",
  "controlled": true,
  "index": 5,
  "lifecycle": "tooltip",
  "size": 7,
  "status": "running",
}
`;

exports[`Joyride > Controlled should have rendered the STEP 7 centered tooltip 1`] = `
{
  "action": "update",
  "controlled": true,
  "index": 6,
  "lifecycle": "ready",
  "size": 7,
  "status": "running",
}
`;
